### ISSUE TITLE
IMAP4Client.lineReceived() fails to parse literals

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2700]
|<img alt="jenner's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jenner reported|
|-|-|
|Trac ID|trac#2700|
|Type|defect|
|Created|2007-06-09 11:38:28Z|

IMAP4Client.lineReceived() parses literals ({<octet>}) only if they are surrounded by whitespaces (like " {10} "), so if the server sends a response and puts a literal within a parenthesized list (e.g. " (({40}\n") the parser fails. The attached patch fixes this issue.

Attachments:

* [lineReceived.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf0/cf06f97a886a3fb6dd08291ca8d0660d8b715b35/76e4937639223f4f3b429e29b0a970d10d67cfa0.patch) (767 bytes) - added by jenner on 2007-06-09 11:49:48Z - A patch for IMAP4Client.lineReceived against trunk (revision: revision: source:/trunk/mail/imap4.py#18812)
* [lineReceived.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf0/cf06f97a886a3fb6dd08291ca8d0660d8b715b35/f99179e99b5fa2b5d4ebf99995cb665250645a15.patch) (767 bytes) - added by jenner on 2007-06-09 11:50:40Z - A patch for IMAP4Client.lineReceived against trunk (revision: revision: source:/trunk/twisted/mail/imap4.py#18812)

<details><summary>Searchable metadata</summary>

```
trac-id__2700 2700
type__defect defect
reporter__jenner jenner
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__mail_imap mail imap
time__1181389108000000 1181389108000000
changetime__1181570904000000 1181570904000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Can you attach a transcript of an IMAP4 session which includes a literal like this?

**Comment by twisted-trac** [comment]
|<img alt="jenner's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|jenner commented|
|-|-|

Replying to [exarkun](#note_1):
> Can you attach a transcript of an IMAP4 session which includes a literal like this?

Sure, here's the relevant part of the session (captured using `tcpflow -C port 143`, names and mail adresses changed keeping the original string size):

```
0004 FETCH 1:* (ENVELOPE)

* 1 FETCH (ENVELOPE ("Mon, 30 Oct 2006 08:40:56 -0600" "RE: Inherited property resolution" (({40}
Xxxxxxxxx, Yyyyyyyyy \(US - Glen Mills\) NIL "zzzzzzzzzz" "fooooooo.com")) (({40}
Xxxxxxxxx, Yyyyyyyyy \(US - Glen Mills\) NIL "zzzzzzzzzz" "fooooooo.com")) (("Maven Users List" NIL "users" "maven.apache.org")) ((NIL NIL "Abbb.Cdddddd" "ab.cdef.com")) ((NIL NIL "users" "maven.apache.org")) NIL NIL "<265E9EE670ED404DA5DBE42428204A4427F07D@...>"))
0004 OK Fetch completed.

```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Fixed in nested-imap4-literal-2700

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

Looks good. If you feel it, lineReceived is missing a docstring, and the ```except ValueError``` path is not tested.

Otherwise please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Good points, both.  Added a docstring and a unit test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Great. There a few trailing whitespaces in imap4.py, then you can merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [20526]) Merge nested-imap4-literal-2700

Author: exarkun
Reviewer: therve
Fixes [#2700](https://github.com/twisted/twisted/issues/2700)

Fix IMAP4Client's parsing of literals which are not preceeded
by whitespace.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
